latte/mine/estimation.py

```diff
@@ -0,0 +1,194 @@
+from typing import Optional, Dict, Union, Any, Tuple
+from dataclasses import dataclass
+
+import geoopt
+import torch
+import torch.nn as nn
+import pytorch_lightning as pl
+from pytorch_lightning.loggers import WandbLogger
+
+from latte.mine import mine
+from latte.models import datamodules as dm
+from latte.manifolds import utils as mutils
+
+
+@dataclass
+class MutualInformationEstimationResult:
+    """A dataclass for holding the result of mutual information estimation using MINE
+
+    Members:
+        mutual_information (float): The estimate of the mutual information as calculated on the test dataset
+        validation_mutual_information (float): The estimate of the mutual information
+                                                as calculated on the validation dataset
+        estimator (mine.MINE): The model trained and used for estimation of the mutual information
+        subspace: In case of also estimating the linear subspace capturing the most information, this holds the
+                  k-frame defining the estimated subspace
+    """
+
+    mutual_information: float
+    validation_mutual_information: float
+    estimator: mine.MINE
+    subspace: Optional[geoopt.ManifoldTensor] = None
+
+
+def _construct_mine_model(
+    manifold: bool,
+    x_size: int,
+    z_size: int,
+    network_width: int,
+    subspace_size: Optional[int] = 0,
+    learning_rate: float = 1e-4,
+    manifold_learning_rate: float = 1e-3,
+    alpha: float = 0.01,
+) -> mine.MINE:
+    """A utility function for creating the MINE model used for estimation of mutual information and possibly searching
+    for the optimal linear subspace.
+
+    Args:
+        manifold: Whether the model should also estimate the linear subspace of the space which X lives in of dimension
+                  `subspace_size` capturing the most mutual information about Z
+        x_size: The dimension of X
+        z_size: The dimension of Z
+        network_width: The side of the hidden layers in the statistics network
+        subspace_size: If `manifold` is True, this controls the dimension of the linear subspace in the space of X
+                       over which the mutual information should be optimised
+        learning_rate: The learning_rate of the MINE model
+        manifold_learning_rate: The learning rate for the k-frame manifold optimiser
+        alpha: The bias-compensating alpha value for the MINE estimator
+
+    Returns:
+        The constructed model
+    """
+    if manifold:
+        assert subspace_size != 0, "No target subspace size provided"
+        S = mine.ManifoldStatisticsNetwork(
```

Comment:
Possible todo item: generalizing this so that we can provide a non-hardcoded network for MINE.